torchao/__init__.py

```diff
@@ -20,24 +21,47 @@
 except PackageNotFoundError:
     __version__ = "unknown"  # In case this logic breaks don't break the build
 
+
 logger = logging.getLogger(__name__)
 
+
+def _parse_version(version_string):
+    """
+    Parse version string representing pre-release with -1
+
+    Examples: "2.5.0.dev20240708+cu121" -> [2, 5, -1], "2.5.0" -> [2, 5, 0]
+    """
+    # Check for pre-release indicators
+    is_prerelease = bool(re.search(r"(git|dev)", version_string))
+    match = re.match(r"(\d+)\.(\d+)\.(\d+)", version_string)
+    if match:
+        major, minor, patch = map(int, match.groups())
+        if is_prerelease:
+            patch = -1
+        return [major, minor, patch]
+    else:
+        raise ValueError(f"Invalid version string format: {version_string}")
+
+
 skip_loading_so_files = False
 # if torchao version has "+git", assume it's locally built and we don't know
 #   anything about the PyTorch version used to build it
 # otherwise, assume it's prebuilt by torchao's build scripts and we can make
 #   assumptions about the PyTorch version used to build it.
 if (not "+git" in __version__) and not ("unknown" in __version__):
-    # torchao v0.13.0 is built with PyTorch 2.8.0. We know that torchao .so
-    # files built using PyTorch 2.8.0 are not ABI compatible with PyTorch 2.9+.
-    # The following code skips importing the .so files if PyTorch 2.9+ is
-    # detected, to avoid crashing the Python process with "Aborted (core
+    # We know that torchao .so files built using PyTorch 2.8.0 are not ABI compatible with PyTorch 2.9+. (see #2919)
+    # The following code skips importing the .so files if incompatible torch version is detected,
+    # to avoid crashing the Python process with "Aborted (core
     # dumped)".
-    # TODO(#2901, and before next torchao release): make this generic for
-    # future torchao and torch versions
-    if __version__.startswith("0.13.0") and str(torch.__version__) >= "2.9":
+    # .dev for nightlies since 2.9.0 has not been released
+    if _parse_version(torch.__version__) == _parse_version(
+        "2.9.0.dev"
+    ) and _parse_version(__version__) == _parse_version("0.14.0"):
```

Comment:
why is this only torch 2.9.0 nightly and torchao == 0.14.0 instead of something like torch >= 2.8 and torchao >= 0.13.0?

Reply:
> if we did torch >= x and torchao >= y we can have false positives if torchao version is y+2 but torch is x+1 when y+2 is built against torch x+2 (so not guaranteed compatible with torch x+1)

I think changing to `>=` for both would skip all future inclusions of this? e.g. torchao >= 0.14.0.dev and torch >= 2.9.0.dev (this will be satisfied for all current and future torch and torchao versions)

Reply:
also to clarify, current version for torchao should be `0.14.0.dev` I think, not sure if you intended to check 0.14.0 or not